### DISCUSSION TITLE
Next.js 13 - How to manage dynamic theming per route

### DISCUSSION THREAD

**Discussion by kungfoolabs** [vercel/next.js#48100] [Help]
### Summary

Hi everyone,

I recently started playing with Next.js 13. Although navigating the Client side and Server side components has been an interesting ride, I believe I managed to get the general gist. However, I am struggling to handle a pretty basic situation that would imagine would come up quite frequently: dynamic theming per route

Basically, the theme of my page for `/` and `/item/{id}` should be different. However, my themes are pulled from the database.  

To accomplish this, I need to manage theming for both Client Side & Server side components. In my `layout.tsx` in the root directory, I have added a `ThemeContextProvider` with a default theme. And then i've created a `ThemeLoader` that allows me to change the theme once the theme is pulled for that specific page

layout.tsx
```
    <html>
      <body>
        <ThemeContextProvider>
            {children}
        </ThemeContextProvider>
      </body>
    </html>
```

page.tsx
```
export default async({ params: { id } }: Context) => {
  const scheme = await fetchScheme(id)

  return (
    <div className="bg-[scheme.backgroundColor]">
      <ThemeLoader scheme={scheme} />

   )
}
```

While this works, it leaves unwanted artifacts with streaming rendering. On refresh, the default theme will always show and then transition into page theme only after the Client Component is streamed/loaded. Even if i were to add a `loading.tsx`, this would not change the fact that the rest of the layout will still show the default theme.

How does one suggest we solve this issue in 13? Can I stop the page from loading/streaming until I get the theme? Is this even possible?

**Comment by 808vita** [comment]
try out SSR ?
make use of getServerSideProps on those specific pages -> fetch styling data &  pass that data as server side props -> de-structure that data on the page component and pass it theme loader .

**Comment by TomRaaff** [comment]
I recently solved this issue the following way:

Say we have a Blog route with specific theming, a Profile route with specific theming and the rest of the application.
In `Globals.css` I defined the variables I used for the 'themes' and gave them a default value. For example:

```css
:root {
    --primary-1: {some light blue}
    --primary-2: {regular blue}
    --primary-3: {dark blue}
}
```

And I defined a couple of css classes that overwrote the theme variables:

```css
.regular-theme {
    --primary-1: {some light blue}
    --primary-2: {regular blue}
    --primary-3: {dark blue}
}

.blog-theme {
    --primary-1: {some light yellow}
    --primary-2: {regular yellow}
    --primary-3: {dark yellow}
}

.profile-theme {
    --primary-1: {some light purple}
    --primary-2: {regular purple}
    --primary-3: {dark purple}
}
```

Then, in the `Layout.tsx` files of each specific route, I added those classes to the top element, which overwrote the theme variables and applied the theming to all underlying elements. Example:

```js
export default async function BlogLayout({ children }) {
  return (
    <div className='blog-theme'>
      {children}
      ...
    </div>
  );
}
```

If you then use those themed css variables in your styling, they will be updated according to the route:
```js
export function Button({children}) {
    return (
        <button style={{ background-color: var(--primary-2); }}>{children}</button>
    )
}
```
